File: app/src/main/java/me/rei_m/hyakuninisshu/presentation/karuta/widget/view/KarutaExamResultView.java
/*
 * Copyright (c) 2017. Rei Matsushita
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
 * the License for the specific language governing permissions and limitations under the License.
 */

package me.rei_m.hyakuninisshu.presentation.karuta.widget.view;

import android.content.Context;
import android.os.Build;
import android.support.annotation.NonNull;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import me.rei_m.hyakuninisshu.domain.model.karuta.Karuta;
import me.rei_m.hyakuninisshu.domain.model.karuta.KarutaIdentifier;
import me.rei_m.hyakuninisshu.event.EventObservable;
import me.rei_m.hyakuninisshu.presentation.util.ViewUtil;

public class KarutaExamResultView extends LinearLayout {

    private static final int NUMBER_OF_KARUTA_PER_ROW = 5;

    private static final int NUMBER_OF_KARUTA_ROW = Karuta.NUMBER_OF_KARUTA / NUMBER_OF_KARUTA_PER_ROW;

    private static final int[] cellViewIdList = new int[Karuta.NUMBER_OF_KARUTA];

    static {
        for (int i = 0; i < Karuta.NUMBER_OF_KARUTA; i++) {
            int viewId;
            if (Build.VERSION.SDK_INT > Build.VERSION_CODES.JELLY_BEAN) {
                viewId = View.generateViewId();
            } else {
                viewId = ViewUtil.generateViewId();
            }
            cellViewIdList[i] = viewId;
        }
    }

    public EventObservable<KarutaIdentifier> onClickKarutaEvent = EventObservable.create();

    public KarutaExamResultView(@NonNull Context context) {
        super(context);
        initialize(context);
    }

    public KarutaExamResultView(@NonNull Context context, @NonNull AttributeSet attrs) {
        super(context, attrs);
        initialize(context);
    }

    public KarutaExamResultView(@NonNull Context context, @NonNull AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initialize(context);
    }

    private void initialize(@NonNull Context context) {
        setOrientation(VERTICAL);
        for (int i = 0; i < NUMBER_OF_KARUTA_ROW; i++) {
            addView(createRow(context, i));
        }
    }

    private LinearLayout createRow(@NonNull Context context, int rowIndex) {

        LinearLayout linearLayout = new LinearLayout(context);
        linearLayout.setOrientation(HORIZONTAL);

        LinearLayout.LayoutParams layoutParams =
                new LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1);

        final int currentRowTopIndex = (rowIndex * NUMBER_OF_KARUTA_PER_ROW);

        for (int i = 0; i < NUMBER_OF_KARUTA_PER_ROW; i++) {
            int totalIndex = currentRowTopIndex + i;
            KarutaExamResultCellView cellView = new KarutaExamResultCellView(context);
            //noinspection ResourceType
            cellView.setId(cellViewIdList[totalIndex]);
            cellView.setGravity(Gravity.CENTER);
            linearLayout.addView(cellView, layoutParams);
        }

        return linearLayout;
    }

    public void setResult(@NonNull boolean[] karutaQuizResultList) {
        for (int i = 0; i < karutaQuizResultList.length; i++) {
            KarutaExamResultCellView cellView = findViewById(cellViewIdList[i]);
            int karutaNo = i + 1;
            cellView.setResult(karutaNo, karutaQuizResultList[i]);
            cellView.setOnClickListener(view -> onClickKarutaEvent.onNext(new KarutaIdentifier(karutaNo)));
        }
    }
}

File: feature/exam/src/main/java/me/rei_m/hyakuninisshu/feature/exam/ui/widget/KarutaExamResultView.kt
/*
 * Copyright (c) 2018. Rei Matsushita
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
 * the License for the specific language governing permissions and limitations under the License.
 */

/* ktlint-disable package-name */
package me.rei_m.hyakuninisshu.feature.exam.ui.widget

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import me.rei_m.hyakuninisshu.domain.model.karuta.Karuta
import me.rei_m.hyakuninisshu.domain.model.karuta.KarutaIdentifier
import me.rei_m.hyakuninisshu.domain.model.quiz.KarutaQuizJudgement
import me.rei_m.hyakuninisshu.feature.corecomponent.flux.Event
import me.rei_m.hyakuninisshu.feature.corecomponent.widget.view.ViewIdHolder

class KarutaExamResultView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    init {
        initialize(context)
    }

    private val _onClickKarutaEvent = MutableLiveData<Event<KarutaIdentifier>>()
    val onClickKarutaEvent: LiveData<Event<KarutaIdentifier>> = _onClickKarutaEvent

    private fun initialize(context: Context) {
        orientation = LinearLayout.VERTICAL
        for (i in 0 until NUMBER_OF_KARUTA_ROW) {
            addView(createRow(context, i))
        }
    }

    private fun createRow(context: Context, rowIndex: Int): LinearLayout {

        val linearLayout = LinearLayout(context).apply {
            orientation = LinearLayout.HORIZONTAL
        }

        val layoutParams = LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f)

        val currentRowTopIndex = rowIndex * NUMBER_OF_KARUTA_PER_ROW

        for (i in 0 until NUMBER_OF_KARUTA_PER_ROW) {
            val totalIndex = currentRowTopIndex + i
            val cellView = KarutaExamResultCellView(context).apply {
                id = cellViewIdList[totalIndex]
                gravity = Gravity.CENTER
            }
            linearLayout.addView(cellView, layoutParams)
        }

        return linearLayout
    }

    fun setResult(judgements: List<KarutaQuizJudgement>) {
        judgements.forEachIndexed { index, karutaQuizJudgement ->
            val (karutaId, isCorrect) = karutaQuizJudgement
            with(findViewById<KarutaExamResultCellView>(cellViewIdList[index])) {
                setResult(karutaId.value, isCorrect)
                setOnClickListener {
                    _onClickKarutaEvent.value =
                        Event(karutaId)
                }
            }
        }
    }

    companion object {

        private const val NUMBER_OF_KARUTA_PER_ROW = 5

        private const val NUMBER_OF_KARUTA_ROW = Karuta.NUMBER_OF_KARUTA / NUMBER_OF_KARUTA_PER_ROW

        private val cellViewIdList = IntArray(Karuta.NUMBER_OF_KARUTA)

        init {
            for (i in 0 until Karuta.NUMBER_OF_KARUTA) {
                val viewId: Int = if (Build.VERSION.SDK_INT > Build.VERSION_CODES.JELLY_BEAN) {
                    View.generateViewId()
                } else {
                    ViewIdHolder.generateViewId()
                }
                cellViewIdList[i] = viewId
            }
        }
    }
}
